backend/doc/grouter.1
.TH "grouter" 1 "22 Septemper 2005" Linux "GINI Components"

.SH NAME
grouter \- The GINI router for creating GINI micro internets.

.SH SYNOPSIS
.B grouter
.BR  -h " | " -V

.B router
.RB "[ " -v
.I level
.RB "] [ " -d 
.I level
.RB "] [ " -i
.I cli
.RB "] [ " -c 
.I config-file
]
.I router-name

.SH DESCRIPTION
This is the router component from the GINI toolkit. It has all the
routing functionalities of a basic linux machine. As its network stack
is restricted upto layer 3, it is much light weighted than using
another UML as a router.

The router can be launched both in interactive and non-interactive
mode. The interactive mode enables interacting with the router with
router configuration commands and monitoring its activity in real
time. It can not be run as a deamon; therefore, when it is started in
a non-interactive mode, it has to be put in the background using '&'
to get the shell prompt back. If launched in the interactive mode,
.B kill
command has to be used to terminate the router.

Similar to 
.BR uswitch ,
.B grouter
also uses the UNIX sockets to emulate the physical links. When a UNIX
socket filename is specified for its physical link, the
.B grouter
creates the specified UNIX socket if it does not already
exist. Therefore, when a GINI network that consists of all switches,
routeres, and UMLs is instantiated, the routers have to started after
the switches but before the UMLs.

.SH OPTIONS

.B -h, --help
.RS 
Prints the usage information and quits.
.RE

.B -V, --version
.RS
Prints the version information and quits.
.RE

.BI "-v, --verbose= " level
.RS
Sets the verbose level to
.IR level .
.RE

.BI "-d, --debug= " level
.RS
Sets the debug level to 
.IR level .
.RE

.BI "-i, --interactive= " cli
.RS 
Sets the interactive level. 
.I cli
can be either 1 or 0, 1 for enabling interactive mode and 0 to disable
it. By default, the interactive mode is disabled. When run in
interactive mode, router will accept commands from
.IR stdin .
GINI router commands can be issued in interactive mode (see the 
.B COMMANDS
section). When run in a non-interactive mode, the router
configurations (or commands) has to be passed through a configuration
file using the
.B -c
option. In the non-interactive mode, the router never exits (even
ctrl-c is disabled) and, to get the shell prompt, the router has to be
launched as a background process using '&' (ctrl-z is disabled). The
router launched in the non-interactive mode can be termintated only
using
.BR kill .
.RE

.BI "-c, --config= " config-file
.RS
When this option is used, the commands in the specified
.I config-file
are executed after the router is started. This option is specifically
useful when the router is launched in the non-interactive mode.
.RE

.BI "-n, --name= " router-name
.RS
Specifies the name of the router. A file named
.I router-name.pid
will be created in the current directory that stores the pid of the
router process. When launched non-interactively, this pid file can be
used to terminate the router as
.RS
.BR "kill " ` cat
.IR router-name.pid `
.RE
The pid file is created at the current directory by default. You can
include a pathname in the
.I router-name
to force the pid file to be created in a different directory.
.RE

.SH COMMANDS
The GINI router accept number of commands to manage and monitoro the
functionality of the router. When run in the interactive modes these
commands can be given via the 
.IR stdin . 
In the interactive mode, the 
.B source
command can be used to feed commands from a batch command file.  When
launched in an interactive mode, the commands can be fed through a
config file using
.B -c 
option.

.B set
.RS
.B set verbose
[
.I level
]
.RS
set the verbose level to
.I level 
(0 < 
.I level
< 7). When
.I level
is omitted, this command will show the current verbose level.
.RE
.RE

.B ifconfig
.RS
.B ifconfig show
[
.I brief
|
.I verbose
]
.RS
Shows the configurations of the interface
.I if 
in brief or verbose mode.
.RE
.BI "ifconfig add " if " -socket " socket-file " -addr " IP-addr " -network " IP-network " -hwaddr " MAC
[
.BI "-gateway " GW
] [
.BI "-mtu " N
]
.RS
Configures and activates the interface
.IR if .
The interface is connected to the switch keeping the socket file
.I socket-file
open. Therefore, if no switch with the specified file is available
this command will have no effect. The interface is assigned with the
6 byte hardware address
.IR MAC ,
the IP address of 
.IR IP-addr ,
and the network address of
.IR IP-network .
Optionally the gateway and mtu of the interface can be set.
.RE
.BI "ifconfig del " if
.RS
Removes the configurations of the interface
.IR if .
.RE
.BI "ifconfig up " if
.RS
If the interface
.I if
is down, it activates it.
.RE
.BI "ifconfig down " if
.RS
Deactivates the interface
.IR if .
The configurations are preserved so that the interface can be
activated just by using the 
.B ifconfig up
command.
.IR if .
.RE
.B ifconfig mod if
(
.BI "-gateway " GW
|
.BI "-mtu " N
)
.RS
Modifies the optional configuration of the interface
.IR if .
.RE
.RE

.B ping 
[
.B -n
]
.I IP-addr
.RS
Similar to a linux box ping command. But, by default, sends only one
packet.
.B -n 
can be used to specify the number of packets to sent.
.RE

.B mtu
.RS
Shows the MTU table.
.RE

.BI "source " config-file
.RS
Similar to the linux source command. Batch executes the commands from
the file
.IR config-file .
.RE

.B halt
.RS
Halts the router.
.RE

.BI "help " command
.RS
Shows a short usage information on the command
.IR command .
.RE

.B route
.RS
.B route show
.RS
Displays the current setup of the routing table.
.RE
.B "route add -dev " if " -net " nw-addr " -netmask " mask
[
.BI "-gw " gw-addr
]
.RS
Adds a routing rule for the interface
.I if
so that all the traffic for the network with network address
.I nw-addr 
with netmask
.I mask
is directed to the gateway
.IR gw-addr .
.RE
.BI "route del " route-number
.RS
Delete the route rule shown in the line number
.I route-number
in the routing table.
.RE
.RE

.B arp
.RS
.B arp show
.RS
Displays the ARP table.
.RE
.BI "arp show -ip " IP-addr
.RS
Shows the ARP entry for the IP address
.IR IP-addr .
.RE
.B arp del
.RS
Deletes all the entries in the ARP table.
.RE
.B "arp del -ip " IP-addr
.RS
Deletes the ARP entry for the IP address
.IR IP-addr .
.RE
.RE

.B version
.RS
Shows the version information of the router.
.RE

.SH BUGS
Report bugs to maheswar@cs.mcgill.ca

.SH AUTHORS
The initial code (named as
.BR uml_router )
is written by Miss. Weiling Xu.

Dr. Muthucumaru Maheswaran did a major restructure of the code and
added many new features and functionalities.

.SH "SEE ALSO"
.BR ifconfig (8),
.BR route (8),
.BR arp (8),
.BR ping (8),
.BR socket (2),
.BR socket (7),
.BR unix (7)


ChangeLog
GINI Toolkit ChangeLog

* gBuilder:
- Fixed placing Mobiles on Canvas

gini-2.0.1
* gBuilder:
- Added button to remove entries from server list in Configuration window
- Removed pyreadline requirement
- Changed item list into debug window
- Made debug window accessible by pressing F10
- Removed subnet/mask properties from Switch
- Improved error handling in general
- Removed arrange toolbar button
- Fixed undroppable canvas in PyQt-4.6
- Fixed crash when placing items on Canvas using PyQt-4.6 on Windows 7

* gRouter:
- Fixed incorrect packet size display on wireshark output
- Integrated patch for gRouter to work on 64-bit machines


gini-2.0.0
- Initial release


backend/src/grouter/grouter.c
/*
 * router.c (GINI router)
 */

//#include <mcheck.h>
#include <slack/std.h>
#include <slack/err.h>
#include <slack/prog.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>
#include <signal.h>
#include <errno.h>
#include <pthread.h>

#include "ip.h"
#include "arp.h"
#include "cli.h"
#include "gnet.h"
#include "packetcore.h"
#include "classifier.h"
#include "filter.h"
#include "openflow_ctrl_iface.h"
#include "openflow_pkt_proc.h"

router_config rconfig = {.router_name=NULL, .gini_home=NULL, .cli_flag=0, .config_file=NULL, .config_dir=NULL, .openflow=0, .ghandler=0, .clihandler= 0, .scheduler=0, .worker=0, .openflow_worker=0, .openflow_controller_iface=0, .openflow_flowtable_timeout=0, .schedcycle=10000};
pktcore_t *pcore;
classlist_t *classifier;
filtertab_t *filter;


Option grouter_optab[] =
{
	{
		"interactive", 'i', "0 or 1", "CLI on for interactive mode (daemon otherwise)",
		required_argument, OPT_INTEGER, OPT_VARIABLE, &(rconfig.cli_flag)
	},
	{
		"config", 'c', "path", "Specify the configuration file",
		optional_argument, OPT_STRING, OPT_VARIABLE, &(rconfig.config_file)
	},
	{
		"confpath", 'p', "path", "Specify directory with configuration files",
		required_argument, OPT_STRING, OPT_VARIABLE, &(rconfig.config_dir)
	},
	{
		"openflow", 'o', "port", "The port number of the OpenFlow controller;"
		" when specified, grouter functions as an OpenFlow 1.0 switch",
		optional_argument, OPT_INTEGER, OPT_VARIABLE, &(rconfig.openflow)
	},
	{
		NULL, '\0', NULL, NULL, 0, 0, 0, NULL
	}
};

Options options[1] = {{ prog_options_table, grouter_optab }};

void setupProgram(int ac, char *av[]);
void removePIDFile();
int makePIDFile(char *rname, char rpath[]);
void shutdownRouter();
int isPIDAlive(int pid);


int main(int ac, char *av[])
{
	char rpath[MAX_NAME_LEN];
	int status, *jstatus;
	simplequeue_t *outputQ, *workQ, *openflowWorkQ, *qtoa;

	// setup the program properties
	setupProgram(ac, av);
	// creates a PID file under router_name.pid in the current directory
	status = makePIDFile(rconfig.router_name, rpath);
	// shutdown the router on receiving SIGUSR1 or SIGUSR2
	redefineSignalHandler(SIGUSR1, shutdownRouter);
	redefineSignalHandler(SIGUSR2, shutdownRouter);

	outputQ = createSimpleQueue("outputQueue", INFINITE_Q_SIZE, 0, 1);
	workQ = createSimpleQueue("work Queue", INFINITE_Q_SIZE, 0, 1);
	if (rconfig.openflow) {
		openflowWorkQ = createSimpleQueue("Work queue for OpenFlow",
			INFINITE_Q_SIZE, 0, 1);
	}

	GNETInit(&(rconfig.ghandler), rconfig.config_dir, rconfig.router_name, outputQ);
	ARPInit();
	IPInit();

	classifier = createClassifier();
	filter = createFilter(classifier, 0);

	pcore = createPacketCore(rconfig.router_name, outputQ, workQ,
							 openflowWorkQ);

	// add a default Queue.. the createClassifier has already added a rule with "default" tag
	// char *qname, char *dqisc, double qweight, double delay_us, int nslots);
	addPktCoreQueue(pcore, "default", "taildrop", 1.0, 2.0, 0);
	rconfig.scheduler = PktCoreSchedulerInit(pcore);
	rconfig.worker = PktCoreWorkerInit(pcore);

	// Initialize OpenFlow packet processor
	if (rconfig.openflow) {
		rconfig.openflow_worker = PktCoreOpenflowWorkerInit(pcore);
	}

	infoInit(rconfig.config_dir, rconfig.router_name);
	addTarget("Output Queue", outputQ);
	qtoa = getCoreQueue(pcore, "default");
	if (qtoa != NULL)
		addTarget("Default Queue", qtoa);
	else
		printf("Error .. found null queue for default\n");

	// Initialize OpenFlow controller interface
	if (rconfig.openflow) {
		rconfig.openflow_controller_iface = openflow_ctrl_iface_init(
				rconfig.openflow);
		rconfig.openflow_flowtable_timeout = openflow_flowtable_timeout_init();
	}

	// start the CLI..
	CLIInit(&(rconfig));


	wait4thread(rconfig.scheduler);
	wait4thread(rconfig.worker);
	if (rconfig.openflow) {
		wait4thread(rconfig.openflow_flowtable_timeout);
		wait4thread(rconfig.openflow_controller_iface);
		wait4thread(rconfig.openflow_worker);
	}
	wait4thread(rconfig.ghandler);
}


void wait4thread(pthread_t threadid)
{
	int *jstatus;
	if (threadid > 0)
		pthread_join(threadid, (void **)&jstatus);
}


void shutdownRouter()
{
	verbose(1, "[main]:: shutting down the GNET handler...");
	GNETHalt(rconfig.ghandler);
	verbose(1, "[main]:: shutting down the packet core... "); fflush(stdout);
	pthread_cancel(rconfig.scheduler);
	pthread_cancel(rconfig.worker);
	if (rconfig.openflow) {
		pthread_cancel(rconfig.openflow_worker);
	}
	verbose(1, "[main]:: shutting down the CLI handler.. ");
	pthread_cancel(rconfig.clihandler);

	// we should cancel CLI thread too??
	verbose(1, "[main]:: removing the PID files... ");
	removePIDFile();
}


void setupProgram(int ac, char *av[])
{
	int indx;

	prog_init();
	prog_set_syntax("[options] router_name");
	prog_set_options(options);
	prog_set_version("2.1");
	prog_set_date("20091024");
	prog_set_author("Muthucumaru Maheswaran <maheswar@cs.mcgill.ca>");
	prog_set_contact(prog_author());
	prog_set_url("http://www.cs.mcgill.ca/~anrl/gini/");
	prog_set_desc("GINI router provides a user-space IP router for teaching and learning purposes.");

	prog_set_verbosity_level(2);

	indx = prog_opt_process(ac, av);

	if (indx < ac)
		rconfig.router_name = strdup(av[indx]);

	if (rconfig.router_name == NULL)
	{
		prog_usage_msg("\n[setupProgram]:: Router name missing.. \n\n");
		exit(1);
	}
	prog_set_name(rconfig.router_name);
	rconfig.gini_home = getenv("GINI_HOME");
	if (rconfig.gini_home == NULL)
	{
		verbose(2, "\n[setupProgram]:: Environment variable GINI_HOME is not set..\n\n");
		exit(1);
	}

}




int makePIDFile(char *rname, char rpath[])
{
	FILE *fp;
	int pid;

	// initialize the config directory..
	if (rconfig.config_dir == NULL)
		rconfig.config_dir = strdup(".");
	sprintf(rpath, "%s/%s.pid", rconfig.config_dir, rname);

	if ((fp = fopen(rpath, "r")) != NULL)
	{
		// another router could be running or left without cleanup!!
		fscanf(fp, "%d", &pid);
		if (isPIDAlive(pid))
		{
			fatal("[makePIDFile]:: Another router is still running under: %s \n", rpath);
			exit(1);
		} else
		{
			remove(rpath);
		}
	}


	if ((fp = fopen(rpath, "w")) == NULL)
	{
		fatal("[makePIDFile]:: ERROR!! unable to create PID file...%s", rpath);
		exit(1);
	}

	fprintf(fp, "%d\n", getpid());
	fclose(fp);

	return TRUE;
}


void removePIDFile()
{
	char rpath[MAX_NAME_LEN];

	sprintf(rpath, "%s/%s.pid", rconfig.config_dir, prog_name());
	remove(rpath);
}


int isPIDAlive(int pid)
{
	if (kill(pid, 0) < 0)
	{
		if (errno == ESRCH)
			return FALSE;

		// this return should never happen..
		return FALSE;
	}
	return TRUE;
}
